Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift

```diff
@@ -60,58 +86,26 @@ public struct AbstractInterpreter {
         var typeChanges = [(Variable, Type)]()
         switch instr.op {
         case is BeginAnyFunctionDefinition:
-            stack.append(currentState)
+            stateDiffsStack.append(VariableMap())
         case is EndAnyFunctionDefinition:
-            let functionState = stack.removeLast()
-            let previousState = stack.removeLast()
-            stack.append(merge(functionState, previousState))
-            typeChanges += getStateChanges(oldState: functionState, newState: stack.last!)
+            let functionStateDiff = stateDiffsStack.removeLast()
+            updateStateDiffs(lastStateDiff: functionStateDiff, typeChanges: &typeChanges)
```

Comment:
And this can probably be called `mergeStates` or so, and there could also be a `resetState(to:)` helper function maybe?
I guess overall, it would be nice if the implementation of the state tracking as diffs and the complete `currentState` was somewhat invisible in this code, so that this would just read as if ever entry on the stack was complete. I think that should be possible even with the diffs. The way I see it, the diffs are just an optimization. Every entry on the stack still represents a complete execution state, just its storage is optimized to only store what has changed. WDYT?

Reply:
so we should create `InterpreterState` struct?

Reply:
Yeah that could be a nice way to implement it!